File: step_counter.cpp
// step_counter.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <json/json.h> // using jsoncpp

using namespace std;

const string DATA_FILE = "steps.json";
const int DEFAULT_GOAL = 10000;

struct HistoryEntry {
    string date;
    int steps;
};

struct Data {
    int goal = DEFAULT_GOAL;
    vector<HistoryEntry> history;
};

class Counter {
private:
    Data data;

    void load() {
        ifstream ifs(DATA_FILE);
        if (!ifs) return;
        Json::Value root;
        ifs >> root;
        data.goal = root.get("goal", DEFAULT_GOAL).asInt();
        for (const auto& item : root["history"]) {
            HistoryEntry e;
            e.date = item["date"].asString();
            e.steps = item["steps"].asInt();
            data.history.push_back(e);
        }
    }

    void save() {
        Json::Value root;
        root["goal"] = data.goal;
        for (const auto& e : data.history) {
            Json::Value item;
            item["date"] = e.date;
            item["steps"] = e.steps;
            root["history"].append(item);
        }
        ofstream ofs(DATA_FILE);
        ofs << root.toStyledString();
    }

    string today() {
        time_t t = time(nullptr);
        tm* now = localtime(&t);
        char buf[11];
        strftime(buf, sizeof(buf), "%Y-%m-%d", now);
        return string(buf);
    }

    HistoryEntry* getToday() {
        string td = today();
        for (auto& e : data.history) {
            if (e.date == td) return &e;
        }
        return nullptr;
    }

public:
    Counter() { load(); }

    void addSteps(int steps) {
        auto entry = getToday();
        if (entry) {
            entry->steps += steps;
        } else {
            HistoryEntry e;
            e.date = today();
            e.steps = steps;
            data.history.push_back(e);
        }
        save();
        int total = entry ? entry->steps : steps;
        cout << "\033[32mДобавлено " << steps << " шагов. Всего сегодня: " << total << "\033[0m" << endl;
    }

    void show() {
        auto entry = getToday();
        int stepsToday = entry ? entry->steps : 0;
        int progress = stepsToday * 100 / data.goal;
        if (progress > 100) progress = 100;
        int barLen = 20;
        int filled = progress * barLen / 100;
        string bar(filled, '█');
        bar.append(barLen - filled, '░');
        cout << "\033[36m📊 Сегодня: " << stepsToday << " шагов\033[0m" << endl;
        cout << "\033[33mЦель: " << data.goal << " шагов\033[0m" << endl;
        cout << "Прогресс: \033[32m" << bar << "\033[0m " << progress << "%" << endl;
        cout << "\033[35mИстория за 7 дней:\033[0m" << endl;
        time_t t = time(nullptr);
        tm* now = localtime(&t);
        for (int i = 6; i >= 0; --i) {
            tm d = *now;
            d.tm_mday -= i;
            mktime(&d);
            char buf[11];
            strftime(buf, sizeof(buf), "%Y-%m-%d", &d);
            string dateStr(buf);
            int steps = 0;
            for (const auto& e : data.history) {
                if (e.date == dateStr) { steps = e.steps; break; }
            }
            cout << "  " << dateStr << ": " << steps << endl;
        }
    }

    void setGoal(int goal) {
        data.goal = goal;
        save();
        cout << "\033[32mЦель установлена: " << goal << " шагов\033[0m" << endl;
    }

    void history() {
        cout << "\033[36mИстория (все записи):\033[0m" << endl;
        sort(data.history.begin(), data.history.end(), [](const HistoryEntry& a, const HistoryEntry& b) {
            return a.date < b.date;
        });
        for (const auto& e : data.history) {
            cout << "  " << e.date << ": " << e.steps << endl;
        }
    }

    void reset() {
        data.history.clear();
        save();
        cout << "\033[32mВсе данные сброшены.\033[0m" << endl;
    }

    void exportCSV(const string& filename) {
        ofstream ofs(filename);
        ofs << "date,steps\n";
        for (const auto& e : data.history) {
            ofs << e.date << "," << e.steps << "\n";
        }
        cout << "\033[32mЭкспортировано в " << filename << " (CSV)\033[0m" << endl;
    }
};

int main(int argc, char* argv[]) {
    int add = 0, goal = 0;
    bool show = false, history = false, reset = false;
    string exportCsv;

    for (int i = 1; i < argc; ++i) {
        string arg = argv[i];
        if (arg == "--add" && i+1 < argc) add = stoi(argv[++i]);
        else if (arg == "--show") show = true;
        else if (arg == "--goal" && i+1 < argc) goal = stoi(argv[++i]);
        else if (arg == "--history") history = true;
        else if (arg == "--reset") reset = true;
        else if (arg == "--export-csv" && i+1 < argc) exportCsv = argv[++i];
    }

    Counter counter;
    if (add > 0) counter.addSteps(add);
    else if (show) counter.show();
    else if (goal > 0) counter.setGoal(goal);
    else if (history) counter.history();
    else if (reset) counter.reset();
    else if (!exportCsv.empty()) counter.exportCSV(exportCsv);
    else cout << "Используйте --help для справки." << endl;
    return 0;
}

File: step_counter.py

#!/usr/bin/env python3
# step_counter.py
import argparse
import json
import os
import sys
from datetime import datetime, timedelta
from colorama import init, Fore, Style

init(autoreset=True)

DATA_FILE = "steps.json"
DEFAULT_GOAL = 10000

class StepCounter:
    def __init__(self):
        self.data = self.load()
        self.goal = self.data.get("goal", DEFAULT_GOAL)
        self.history = self.data.get("history", [])

    def load(self):
        if os.path.exists(DATA_FILE):
            try:
                with open(DATA_FILE, 'r') as f:
                    return json.load(f)
            except:
                pass
        return {"goal": DEFAULT_GOAL, "history": []}

    def save(self):
        with open(DATA_FILE, 'w') as f:
            json.dump({"goal": self.goal, "history": self.history}, f, indent=2)

    def get_today(self):
        today = datetime.now().strftime("%Y-%m-%d")
        for entry in self.history:
            if entry["date"] == today:
                return entry
        return None

    def add_steps(self, steps):
        today = datetime.now().strftime("%Y-%m-%d")
        entry = self.get_today()
        if entry:
            entry["steps"] += steps
        else:
            self.history.append({"date": today, "steps": steps})
        self.save()
        print(Fore.GREEN + f"Добавлено {steps} шагов. Всего сегодня: {entry['steps'] if entry else steps}")

    def show(self):
        today = datetime.now().strftime("%Y-%m-%d")
        entry = self.get_today()
        steps_today = entry["steps"] if entry else 0
        progress = min(100, int((steps_today / self.goal) * 100))
        bar_len = 20
        filled = int(bar_len * progress / 100)
        bar = "█" * filled + "░" * (bar_len - filled)
        print(Fore.CYAN + f"📊 Сегодня: {steps_today} шагов")
        print(Fore.YELLOW + f"Цель: {self.goal} шагов")
        print(f"Прогресс: {Fore.GREEN}{bar} {progress}%")
        # История за 7 дней
        print(Fore.MAGENTA + "\nИстория за 7 дней:")
        today_dt = datetime.now()
        for i in range(6, -1, -1):
            d = (today_dt - timedelta(days=i)).strftime("%Y-%m-%d")
            found = False
            for entry in self.history:
                if entry["date"] == d:
                    print(f"  {d}: {entry['steps']}")
                    found = True
                    break
            if not found:
                print(f"  {d}: 0")

    def set_goal(self, goal):
        self.goal = goal
        self.save()
        print(Fore.GREEN + f"Цель установлена: {goal} шагов")

    def history(self):
        print(Fore.CYAN + "История (все записи):")
        for entry in sorted(self.history, key=lambda x: x["date"]):
            print(f"  {entry['date']}: {entry['steps']}")

    def reset(self):
        self.history = []
        self.save()
        print(Fore.GREEN + "Все данные сброшены.")

    def export_csv(self, filename):
        import csv
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(["date", "steps"])
            for entry in self.history:
                writer.writerow([entry["date"], entry["steps"]])
        print(Fore.GREEN + f"Экспортировано в {filename} (CSV)")

def main():
    parser = argparse.ArgumentParser(description="Шагомер (виджет)")
    parser.add_argument("--add", type=int, help="Добавить шаги за сегодня")
    parser.add_argument("--show", action="store_true", help="Показать сегодняшний прогресс и историю")
    parser.add_argument("--goal", type=int, help="Установить дневную цель")
    parser.add_argument("--history", action="store_true", help="Показать всю историю")
    parser.add_argument("--reset", action="store_true", help="Сбросить все данные")
    parser.add_argument("--export-csv", help="Экспорт истории в CSV")
    args = parser.parse_args()

    counter = StepCounter()
    if args.add:
        counter.add_steps(args.add)
    elif args.show:
        counter.show()
    elif args.goal:
        counter.set_goal(args.goal)
    elif args.history:
        counter.history()
    elif args.reset:
        counter.reset()
    elif args.export_csv:
        counter.export_csv(args.export_csv)
    else:
        parser.print_help()

if __name__ == "__main__":
    main()
